TypeReference usage clarification

kenegozi opened the pull request:
This change makes TypeReference usage a little bit easier

Summary of the changes (Less than 80 chars)

- Marking it abstract make it so that usages cannot "forget" subclassing (or omitting the { } part). Previously the error message would be confusing as the supertype would be Object.
- Removing the instanceof check and relying on exception since instantiating TypeReference without a type parameter (new TypeReference() ) *is* exceptional as it goes against the intention of this class altogether.
- Added some clarification to the javadoc regarding intended usage, and a link to prior art for further reading.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=25719) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=25719) before we can accept your contribution.<br/><br/>:x: kenegozi [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=25719)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=25719) it.</sub>

wtgodbe commented on src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/TypeReference.java:
```diff
@@ -3,30 +3,42 @@
+     * This class implements Super Type Tokens (Gafter's Gadget) as a way to get a reference to generic types in 
+     * spite of type erasure since, sadly, {@code Foo<Bar>.class} is not valid Java.
+     * 
+     * To get the Type of Class Foo<Bar>, use the following syntax:
```
Looks like this javadoc syntax is off:

>   F:\workspace\_work\1\s\src\SignalR\clients\java\signalr\core\src\main\java\com\microsoft\signalr\TypeReference.java:24: error: unknown tag: Bar

wtgodbe reviewed (changes requested):
This looks good, thanks! Once the javadoc issue is resolved I'll approve

BrennanConroy commented:
Do we want to move this to RC2?

wtgodbe commented:
> Do we want to move this to RC2?

Good point - I think that'd be good so that the YourPhone team can get this along with the rest of the rc2 changes.

@kenegozi can you retarget this to the `release/5.0-rc2` branch?

kenegozi commented:
no problem. I'll fix the thing and rebase onto rc2

kenegozi committed:
TypeReference usage clarification

This change makes TypeReference usage a little bit easier

- Marking it abstract make it so that usages cannot "forget" subclassing (or omitting the { } part). Previously the error message would be confusing as the supertype would be Object.
- Removing the instanceof check and relying on exception since instantiating TypeReference without a type parameter (new TypeReference() ) *is* exceptional as it goes against the intention of this class altogether.
- Added some clarification to the javadoc regarding intended usage, and a link to prior art for further reading.

kenegozi commented:
@wtgodbe thanks for pointing out the error message. I could not find the actual error message on github's UI nor at the azure pipelines page. Then when running `gradlew javadoc` locally it ran without the doclint option and passed as well. I had to run `javadoc TypeReference.java  -Xdoclint:all` to repro the error message locally.

kenegozi commented:
I see some failures in the CI report but they are at far-away parts of the codebase. Is HEAD on 5.0-rc2 consistently green?

BrennanConroy commented:
No worries, some machines weren't allocated in time and the build timed out. I've respun the timed out builds.

kenegozi commented:
Looks like another run is stuck at 11h of runtime :(

wtgodbe commented:
I just hit retry again - we've unfortunately been hit by some flakiness in machine availability recently 😦

wtgodbe commented:
@Pilchie can we take this for RC2? It improves the docs, usability & correctness of the newly introduced `TypeReference`, which messagePack users can use to convert a Java Class to a Type.

Pilchie commented:
Approved for RC2.

wtgodbe commented:
Thanks @kenegozi!

kenegozi commented:
Do you need me to front-port it to the main development branch?

Pilchie commented:
Nope, we've got a bot that will handle that.  Thanks for the offer though!
